Efficient miner RPC call

gandrewstone opened the pull request:
Submitted on behalf of / in collaboration with @lonoami, since I did some last minute cleanups and python test in my own repo.

lonoami committed:
Added stratum protocol. With tests. (Still need to move the tests to other files.)

lonoami committed:
renamed functions and varaibles to non-stratum names

gandrewstone committed:
add bitcoin-miner example CPU miner.  Document merkle API calls and use them in the new mining block candidate RPC calls

lonoami committed:
Fully implemented Bitcoin Candidate

lonoami committed:
Changed submit candidate to retry sending to bitcoind. Before only once.

gandrewstone committed:
multithread bitcoin-miner, increment candidate ids rather than use a random number, clean up old mining candidates ONLY if they are old

gandrewstone committed:
issue p2pkh coinbase script and fix formatting

gandrewstone committed:
fix merge error

lonoami committed:
Bitcoin mining documentation added

lonoami committed:
bitcoin-miner: Added help message, ability to set blockversion and it now uses its own time

gandrewstone committed:
Merge pull request #20 from lonoami/minerRpc

Miner rpc

gandrewstone committed:
custom allowed_args should be able to be defined in the executable that uses them, not in one spot.  Also clean up markup documentation

gandrewstone committed:
move miner-specific allowed args into bitcoin-miner.  exchange JSON not string-encoded json, and regularize json dictionary names with other RPC calls.  Add python unit test

gandrewstone committed:
pass univalue object not stringified, and properly parse the miningsolution return

lonoami committed:
Changed bitcoin-miner to work with changes in submitminingsolution function. Fixed help messages.

gandrewstone committed:
merge proper error checking in bitcoin-miner

gandrewstone committed:
merge proper error checking in bitcoin-miner

gandrewstone committed:
remove deliberate breaking of the submission which was used to test error handling

gandrewstone committed:
fix mining test python test name

ptschip commented on src/unlimited.cpp:
```diff
@@ -1615,6 +1617,426 @@ UniValue setlog(const UniValue &params, bool fHelp)
+
+/** Stratum Protocol - begin */
+
+
```
Can we move these functions out of unlimited.cpp and create an /rpc/stratum.cpp or something similar?

gandrewstone committed:
add windows resource file

lonoami committed:
Fixed Windows build

lonoami committed:
Fixed 32 bit unsigned int problem in submitminingsolution

lonoami committed:
Changed error handling for: id not found

lonoami committed:
Added more debug logging

sickpig reviewed (commented):
dunno what's happened but `merge proper error checking in bitcoin-miner` has been merged twice. guess you could squash the 2.

then PR need to be rebased

sickpig commented on src/test/merkle_tests.cpp:
```diff
@@ -72,8 +72,57 @@ static inline int ctz(uint32_t i)
+    {
+        // add a random leaf
+        uint256 temp;
+        for (uint j = 0; j < 32; j++)
```
is this portable? AFAIR `uint` is not standard.

sickpig commented on src/unlimited.cpp:
```diff
@@ -40,6 +41,7 @@
 #include "validationinterface.h"
 #include "version.h"
 
+
```
why this new line?

sickpig commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,730 @@
+
+#include <univalue.h>
+
+using namespace std;
```
If memory serves we made a few attempt to avoid using implicit call to `std` library. e.g. see [L41](https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1139/files#diff-c5960742195e7bc64c74b375d392cde3R41)

gandrewstone replied:
We agreed that you must never "using namespace" in header files.  I do not think we agreed for source files.  Code can get extremely ugly without allowing it, so I think we should leave it up to the developer's discretion.

sickpig replied:
didn't recall the decision but it doesn't matter much.

that said what do you mean by ugly?

gandrewstone replied:
for example:
std::chrono::high_resolution_clock::time_point t1 = std::chrono::high_resolution_clock::now();

vs:
using namespace std::chrono::high_resolution_clock;
...
time_point t1 = now();

lonoami replied:
In addition you can use namespaces scoped by { }. This will ensure that the namespace is not in use in the code below the }.

util_tests.cpp:
```
BOOST_AUTO_TEST_CASE(util_Logging)
{
    {
        using namespace Logging; 
```

sickpig commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,730 @@
+std::map<std::pair<void *, void *>, LockStack> lockorders;
+boost::thread_specific_ptr<LockStack> lockstack;
+
+static const int CONTINUE_EXECUTION = -1;
```
nit: mix commenting styles.

sickpig commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,730 @@
+
+    uint32_t startNonce = header.nNonce = std::rand();
+
+    // printf("searching...target: %s\n",arith_uint256().SetCompact(header.nBits).GetHex().c_str());
```
to be removed?

sickpig commented on src/test/merkle_tests.cpp:
```diff
@@ -72,8 +72,57 @@ static inline int ctz(uint32_t i)
+        std::vector<uint256> leaves = leafData;
+        leaves.resize(size);
+
+        // printf("leaves: %u\n", (unsigned int) leaves.size());
```
remove or mv it to a `LOG` statement

gandrewstone committed:
Merge pull request #22 from lonoami/minerRpc

Miner rpc

gandrewstone committed:
fix integer sizes in error printout

gandrewstone committed:
Merge branch 'dev' into minerRpc

gandrewstone committed:
fix miningtest error detection to reflect slightly different way incorrect id is being reported

gandrewstone committed:
merge

gandrewstone committed:
fix comment styles, remove old commented-out printf

ptschip commented on src/test/merkle_tests.cpp:
```diff
@@ -72,8 +72,57 @@ static inline int ctz(uint32_t i)
+{
+    uint256 running = elem;
+    for (auto i : leaves)
+    {
```
use a reference and maybe call it "hash" .   "i" ->>  "&hash"

ptschip commented on src/test/merkle_tests.cpp:
```diff
@@ -72,8 +72,57 @@ static inline int ctz(uint32_t i)
+        leaves.resize(size);
+
+        // printf("leaves: %u\n", (unsigned int) leaves.size());
+        uint256 root = ComputeMerkleRoot(leaves);
```
remove commented out printf.

ptschip commented on src/unlimited.cpp:
```diff
@@ -1615,6 +1618,222 @@ UniValue setlog(const UniValue &params, bool fHelp)
+}
+
+uint256 CalculateMerkleRoot(uint256 &coinbase_hash, std::vector<uint256> merklebranches)
+{
```
pass reference  "std::vector<uint256> &merklebranches"

lonoami replied:
Thanks I missed that.

ptschip commented on src/globals.cpp:
```diff
@@ -276,6 +276,8 @@ CThinBlockData thindata; // Singleton class
 uint256 bitcoinCashForkBlockHash = uint256S("000000000000000000651ef99cb9fcbe0dadde1d424bd9f15ff20136191a5eec");
 
+map<int64_t, CMiningCandidate> miningCandidatesMap;
+
```
-map<int64_t, CMiningCandidate> miningCandidatesMap
+map<int64_t, CMiningCandidate> miningCandidatesMap GUARDED_BY(cs_main);

lonoami replied:
Did not know GUARDED_BY syntax. 
Does it make miningCandidatesMap threadsafe?
If so these can be removed: LOCK(cs_main); Yes/No?

ptschip commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,723 @@
+}
+
+static uint256 CalculateMerkleRoot(uint256 &coinbase_hash, std::vector<uint256> merklebranches)
+{
```
-std::vector<uint256> merklebranches
+std::vector<uint256> &merklebranches

lonoami replied:
Thanks I missed that.

ptschip commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,723 @@
+static bool CpuMineBlockHasher(CBlockHeader *pblock,
+    vector<unsigned char> &coinbaseBytes,
+    std::vector<uint256> merklebranches)
+{
```
-std::vector merklebranches
+std::vector &merklebranches

lonoami replied:
Thanks I missed that.

ptschip commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,723 @@
+// This function returns either one of EXIT_ codes when it's expected to stop the process or
+// CONTINUE_EXECUTION when it's expected to continue further.
+static int AppInitRPC(int argc, char *argv[])
+{
```
Many of these functions are cut and paste duplicates from bitcoin-cli.cpp.  Seems to make better sense to make them non-static in bitcoin-cli.cpp rather than copying them all to here.

AppInitRPC()
HttpReply()
http_request_done()
CallRpc()

ptschip commented on src/bitcoin-miner.cpp:
```diff
@@ -0,0 +1,723 @@
+}
+
+double GetDifficulty(uint64_t nBits)
+{
```
GetDifficulty() is also a function in Blockchain.cpp.  Maybe this one should be made "static"?

ptschip commented on qa/rpc-tests/miningtest.py:
```diff
@@ -0,0 +1,156 @@
+        tx.vout[0].scriptPubKey = CScript([OP_1])
+        #tx.vout.append(CTxOut(5000000000, CScript(["OP_1"])))
+        # tx.vin.append(
+
```
remove commented out code

gandrewstone replied:
thx missed that!

gandrewstone committed:
factor common RPC client code into the RPC client library

ptschip commented:
you still need the Lock(cs_main)...the guarded_by is just making it
easier to know what requires a cs_main lock, and since there are so many
things that need a cs_main lock it's easy for a dev or new dev to lose
track.

On 21/06/2018 11:52 AM, johan wrote:
>
> *@lonoami* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/globals.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1139#discussion_r197242176>:
>
> > @@ -276,6 +276,8 @@ CThinBlockData thindata; // Singleton class
>  
>  uint256 bitcoinCashForkBlockHash = uint256S("000000000000000000651ef99cb9fcbe0dadde1d424bd9f15ff20136191a5eec");
>  
> +map<int64_t, CMiningCandidate> miningCandidatesMap;
> +
>
> Did not know GUARDED_BY syntax.
> Can explicit locks be removed?
> Such as:
> RmOldMiningCandidates()
> {
> LOCK(cs_main);
> ...
> it = miningCandidatesMap.erase(it);
> ...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1139#discussion_r197242176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0LUricSsX_utXKdXsvCi3zRTVPjUks5t--uCgaJpZM4UpfE3>.
>

gandrewstone committed:
fix various small issues like using const references, and GUARDED_BY

jasonbcox commented on doc/miner.md:
If the miner has the merkle root and path to the coinbase, listing all of the branches shouldn't be necessary.

Also, an indication of where the coinbase is located within the merkle tree will make this API more robust against potential changes in the future that modify transaction ordering.  See how we're planning on solving that for ABC: https://docs.google.com/document/d/10YVLdltuppEoizAVoovPTOZbqthbsa2t0MXHhaDo4bI/

gandrewstone replied:
This is just the path to the coinbase.  It contains the merkle root of the all the branches of the tree (subtrees) that the miner does not need.  I think that creating a list of dictionaries to solve the non-0 coinbase is excessively complicated and will increase parsing time.  Instead think about adding a single integer where each bit tells you to do CAT(A,B) or CAT(B,A).  This can easily be added if non-zero coinbase indexes ever happen.

jasonbcox commented on doc/miner.md:
Though not strictly necessary, it may be helpful to define a cookie that gets passed back to the node software as part of submitminingsolution.  This will allow you to extend and improve the API without forcing miners to update their API calls (they would only need to upgrade the node software and it would work out of the box).

gandrewstone replied:
we have a cookie called "id".

gandrewstone committed:
resolve issue where the miner produces bad blocks on testnet because the time requires a difficulty reduction but the block does not reduce difficulty

ptschip committed:
Update windows rebuild-bitcoin.bat script to account for
bitcoin-miner.exe

When finished building all binaries we need to copy the
bitcoin-miner.exe to the correct build output folder along with all the
other binaries.

gandrewstone committed:
Merge pull request #23 from ptschip/dev_minerRpc

Update windows rebuild-bitcoin.bat script to account for

gandrewstone committed:
change merklebranches to merkleProof because the terminology of a merkle branch is confusing

ptschip reviewed (approved):
good idea to change to merkleproof... re-approved
